waterbutler/providers/cloudfiles/provider.py

```
@@ -361,13 +394,110 @@ def _extract_endpoints(self, data):
                             break
 
         return [
-            self._serialize_folder_metadata(item)
+            self._serialize_metadata(item)
             for item in data
         ]
 
-    def _serialize_folder_metadata(self, data):
+    def _serialize_metadata(self, data):
         if data.get('subdir'):
             return CloudFilesFolderMetadata(data)
         elif data['content_type'] == 'application/directory':
             return CloudFilesFolderMetadata({'subdir': data['name'] + '/'})
         return CloudFilesFileMetadata(data)
+
+    @ensure_connection
+    async def create_folder(self, path, **kwargs):
+        """Create a folder in the current provider at `path`. Returns a `BaseFolderMetadata` object
+        if successful.  May throw a 409 Conflict if a directory with the same name already exists.
+        Enpoint information can be found here:
+        https://developer.rackspace.com/docs/cloud-files/v1/general-api-info/pseudo-hierarchical-folders-and-directories/
+        :param path: ( :class:`.WaterButlerPath` )User-supplied path to create. Must be a directory.
+        :param kwargs: dict unused params
+        :rtype: :class:`.BaseFileMetadata`
+        :raises: :class:`.CreateFolderError`
+        """
+
+        resp = await self.make_request(
+            'PUT',
+            functools.partial(self.sign_url, path, 'PUT'),
+            expects=(200, 201),
+            throws=exceptions.CreateFolderError,
+            headers={'Content-Type': 'application/directory'}
+        )
+        await resp.release()
```

What does Rackspace do if there is a conflict? We should return a specific error if possible. 

ie:
```
        async with self.request(
            'POST',
            self.build_url('folders'),
            data={
                'name': path.name,
                'parent': {
                    'id': path.parent.identifier
                }
            },
            expects=(201, 409),
            throws=exceptions.CreateFolderError,
        ) as resp:
            # Catch 409s to avoid race conditions
            if resp.status == 409:
                raise exceptions.FolderNamingConflict(path.name)
```

---

Folder naming conflict are handled in postvalidation. Here: https://github.com/CenterForOpenScience/waterbutler/blob/develop/waterbutler/server/api/v1/provider/create.py#L65

---

👍